modules/caddyhttp/fileserver/browse.go

```diff
@@ -48,9 +48,10 @@ func (fsrv *FileServer) serveBrowse(root, dirPath string, w http.ResponseWriter,
 	// "/b/c" on a path like "/a" end up going to "/b/c" instead
 	// of "/a/b/c" - so we have to redirect in this case
 	if !strings.HasSuffix(r.URL.Path, "/") {
-		fsrv.logger.Debug("redirecting to trailing slash to preserve hrefs", zap.String("request_path", r.URL.Path))
-		r.URL.Path += "/"
-		http.Redirect(w, r, r.URL.String(), http.StatusMovedPermanently)
+		oldReq := r.Context().Value(caddyhttp.OriginalRequestCtxKey).(http.Request)
+
+		fsrv.logger.Debug("redirecting to trailing slash to preserve hrefs", zap.String("request_path", oldReq.URL.Path))
+		http.Redirect(w, r, oldReq.URL.Path+"/", http.StatusMovedPermanently)
```

Comment:
I think we should probably keep the `.String()` call. This makes the redirect go to the fully qualified URL rather than just the absolute path. It'll make sure the path is properly escaped as well, etc.

Reply:
Isn't properly escaping guaranteed by `net/http` using `path.Clean`? I'm mainly doing this just to avoid copying a `url.URL` or mutating `oldReq` itself.

Reply:
Yeah, you're probably right.

Reply:
Redirecting to `oldReq.URL.Path+"/"` will drop the query string and other URL components, though. I think the current way is more correct.

Reply:
Nevermind, `http.Redirect()` apparently reconstructs the full URL given a relative target.

Reply:
I think you're actually right, and I've only realized this just a few hours ago as well. Reading the `http.Redirect` code, I don't think it would reconstruct the URL with the old query given `oldReq.URL.Path+"/"`, so I should've given it `oldReq.URL.String()`.

I can hand in another PR if my guess is right.

Reply:
Could you test it out? And if so, sure, another PR would be appreciated. 👍

Reply:
I think you're right after all. With the path redirection, I'm not getting the query parameters back.

```
―❤―▶cat main.go
...
http.Redirect(w, r, r.URL.Path+"/", 200)
...

―❤―▶ curl -v localhost:8080/path?a=b
...
< HTTP/1.1 200 OK
< Location: /path/
< 
<a href="/path/">OK</a>.
```

With the `String()` redirection, I'm getting the query parameters as expected in the `Location` header.

```
―❤―▶cat main.go
...
r.URL.Path += "/"
http.Redirect(w, r, r.URL.String(), 200)
...

―❤―▶ curl -v localhost:8080/path?a=b
< HTTP/1.1 200 OK
< Location: /path/?a=b
< 
<a href="/path/?a=b">OK</a>.
```

Reply:
I've filed PR #4196 for this.